c++/src/capnp/layout.c++

```diff
@@ -344,7 +345,8 @@ struct WireHelpers {
     //   segment belonging to the arena.  `ref` will be initialized as a non-far pointer, but its
     //   target offset will be set to zero.
 
-    if (orphanArena == nullptr) {
+    // Canonical messages can't be crafted from orphans, don't try
+    if ((orphanArena == nullptr) || canonical) {
```

Comment:
If `orphanArena` is non-null then taking this branch will crash (because `segment` is null).

It doesn't look to me like adding `canonical` as a parameter of this method is accomplishing anything. You throw an exception on one branch, but calling `isCanonical()` later would have detected the problem anyway.

I think it would be best to revert the changes to this method.